EEJT-0002/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java

```diff
@@ -0,0 +1,103 @@
+import com.epam.esm.exceptionhandler.exception.ServerException;
+import com.epam.esm.giftcertficate.model.GiftCertificate;
+import com.epam.esm.giftcertficate.repository.GiftCertificateRepository;
+import com.epam.esm.tag.model.Tag;
+import com.epam.esm.tag.service.TagService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.sql.SQLIntegrityConstraintViolationException;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+@Service
+@EnableTransactionManagement
+public class GiftCertificateService {
+
+    private final GiftCertificateRepository giftCertificateRepository;
+    private final TagService tagService;
+
+    @Autowired
+    public GiftCertificateService(GiftCertificateRepository giftCertificateRepository, TagService tagService) {
+        this.giftCertificateRepository = giftCertificateRepository;
+        this.tagService = tagService;
+    }
+
+
+    public List<GiftCertificate> getAllCertificates() {
+
+        return giftCertificateRepository.getAllGiftCertificates();
+    }
+
+    @Transactional
+    public boolean createCertificate(GiftCertificate giftCertificate) {
+        if (!giftCertificateRepository.isGiftCertificateExist(giftCertificate)) {
+
+            boolean result = giftCertificateRepository.createCertificate(giftCertificate);
+
+            if (giftCertificate.getTags() != null) {
+                tagService.isTagsExistOrElseCreate(giftCertificate.getTags());
+
+                List<Long> listTagsId = tagService.getTagsIdByTags(giftCertificate.getTags());
+                giftCertificateRepository.createGiftCertificateTagRelationship(listTagsId, getGiftCertificateId(giftCertificate));
+            }
+
+            return result;
+
+        } else {
+            throw new ServerException("Such certificate has already existed");
+        }
+
+    }
+
+    public List<GiftCertificate> getCertificateById(long id) {
+        List<GiftCertificate> giftCertificate = giftCertificateRepository.getCertificateById(id);
+        if (!giftCertificate.isEmpty()) {
+            return giftCertificate;
+        }
+        throw new ItemNotFoundException("There are no gift certificate with id= " + id);
+    }
+
+    public boolean deleteGiftCertificate(long id) {
+        if (giftCertificateRepository.deleteCertificate(id)) {
+            return true;
+        }
+        throw new ItemNotFoundException("There is no gift certificate with id= " + id);
+    }
+
+    @Transactional
+    public List<GiftCertificate> updateGiftCertificate(long id, List<Tag> tags, Optional<Map<String, String>> updatesMap) {
+
+        updatesMap.get().remove("tags");
+
+        if (giftCertificateRepository.updateGiftCertificate(id, updatesMap)) {
```

Comment:
Why do you pass Oprional `updatesMap` to repository if you already called `get` earlier?

Reply:
Optional should be used as return type, and not as entry parameter of the method

Reply:
remake